x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java

```diff
@@ -137,6 +190,142 @@ public <T extends TransportResponse> void sendRequest(
+            }
+
+            private <T extends TransportResponse> void sendWithRemoteAccessHeaders(
+                final RemoteAccessCredentials remoteAccessCredentials,
+                final Transport.Connection connection,
+                final String action,
+                final TransportRequest request,
+                final TransportRequestOptions options,
+                final TransportResponseHandler<T> handler
+            ) {
+                final String remoteClusterAlias = remoteAccessCredentials.clusterAlias();
+                if (connection.getVersion().before(VERSION_REMOTE_ACCESS_HEADERS)) {
+                    throw new IllegalArgumentException(
+                        "Settings for remote cluster ["
+                            + remoteClusterAlias
+                            + "] indicate remote access headers should be sent but target cluster version ["
+                            + connection.getVersion()
+                            + "] does not support receiving them"
+                    );
+                }
+
+                logger.debug("Sending request to [{}] with remote access headers for [{}] action", remoteClusterAlias, action);
+
+                final Authentication authentication = securityContext.getAuthentication();
+                assert authentication != null : "authentication must be present in security context";
+
+                final ThreadContext threadContext = securityContext.getThreadContext();
+                final var contextRestoreHandler = new ContextRestoreResponseHandler<>(threadContext.newRestorableContext(true), handler);
+                authzService.retrieveRemoteAccessRoleDescriptorsIntersection(
+                    remoteClusterAlias,
+                    authentication.getEffectiveSubject(),
+                    ActionListener.wrap(roleDescriptorsIntersection -> {
+                        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+                            remoteAccessCredentials.writeToContext(threadContext);
+                            new RemoteAccessAuthentication(authentication, roleDescriptorsIntersection).writeToContext(threadContext);
+                            sender.sendRequest(connection, action, request, options, contextRestoreHandler);
+                        }
+                    }, e -> contextRestoreHandler.handleException(new SendRequestTransportException(connection.getNode(), action, e)))
+                );
+            }
+
+            private boolean isAllowlistedForRemoteAccessHeaders(final TransportRequest request) {
+                return request instanceof ShardSearchRequest
+                    || request instanceof ShardFetchRequest
+                    || request instanceof SearchRequest
+                    || request instanceof ClusterSearchShardsRequest;
+            }
+
+            record RemoteAccessCredentials(String clusterAlias, String credentials) {
```

Comment:
This ended up a pretty neat abstraction (cheers for the suggestion Yang!) -- I think it makes sense to pull this up even further, to `RemoteClusterAuthorizationResolver` but I think that refactor makes more sense in a follow up PR to keep the current PR more focused.